### PR TITLE
Fix stdc atomics detection and add vs2019 msvc job

### PR DESCRIPTION
**Pull request by satk0**
Made a PR as mentioned in issue #586

### CODE REVIEW

**Comment by saghul** [comment]
I think you need to use the windows-2019 image: https://github.com/actions/runner-images

**Comment by satk0** [comment]
Hmm, maybe escaping `/experimental:c11atomics` for not vs2022 in CMakelists.txt would somehow help?

**Comment by saghul** [comment]
I think the problem is that we now unconditionally import the atomics compat header: https://github.com/search?q=repo%3Aquickjs-ng%2Fquickjs%20%23include%20%22quickjs-c-atomics.h%22&type=code

Skipping the run-262 build should be doable, and we only use it in a single place in quickjs-libc.c so I think if we address that we should be good to go.

In your case (radare2) I guess you only care about the former problem, right?

**Comment by saghul** [comment]
Looking at https://learn.microsoft.com/en-us/cpp/overview/compiler-versions?view=msvc-170 and knowing C11 atomics made it to Visual Studio 2022 version 17.5,
any MSVC < 19.35 is currently not supported.

A working build:

```
-- The C compiler identification is MSVC 19.41.34120.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: C:/Program Files/Microsoft Visual Studio/2022/Enterprise/VC/Tools/MSVC/14.41.34120/bin/Hostx64/x64/cl.exe - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- No build type selected, default to Release
-- Building in Release mode
-- Building with MSVC 19.41.34120.0 on Windows-10.0.20348
```

A failed one:

```
-- The C compiler identification is MSVC 19.29.30154.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: C:/Program Files (x86)/Microsoft Visual Studio/2019/Enterprise/VC/Tools/MSVC/14.29.30133/bin/Hostx64/x64/cl.exe - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- No build type selected, default to Release
-- Building in Release mode
-- Building with MSVC 19.29.30154.0 on Windows-10.0.17763
```

I think our best bet here is to use Windows specific code for those older compiler versions. We could borrow some code from here I think: https://github.com/videolan/dav1d/blob/master/include/compat/msvc/stdatomic.h

Pinging @bnoordhuis so he is aware :-)

**Comment by satk0** [comment]
Thanks a ton for such a rich answer, it's amazing how much I can learn here 😀 
> In your case (radare2) I guess you only care about the former problem, right?

Yes, the sole problem with radare2 was that we omitted adding C standard to meson build, so `__STDC_NO_ATOMICS__` was undefined. 

> I think our best bet here is to use Windows specific code for those older compiler versions.

Will `#if __has_include("stdatomic.h")` check help in this case?

**Comment by saghul** [comment]
> Will `#if __has_include("stdatomic.h")` check help in this case?

Alas no, because `__has_include` is not supported in MSVC for C (it might, but just for C++)

**Comment by satk0** [comment]
Meh, could you please give me any hint on how this can be solved 😕 ?

**Comment by bnoordhuis** [comment]
You can probably guard on:
```c
#if defined(_MSC_VER) && _MSC_VER < 1935
// do whatever for old msvc
#else
#include <stdatomic.h>
#endif
```
edit: `s/1930/1935/` for 17.5

**Comment by saghul** [comment]
To add, you can borrow the functions we need from https://github.com/videolan/dav1d/blob/master/include/compat/msvc/stdatomic.h and snadwhich them in the if-defs Ben pointed out.

**Comment by satk0** [comment]
Great! Thank you both for your answers, will try to copy it here

**Comment by satk0** [comment]
still failing :/

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -217,6 +217,31 @@ jobs:
...
+      fail-fast: false
+      matrix:
+        arch: [x64, Win32]
+        buildType: [Debug, Release]
+    steps:
+      - uses: actions/checkout@v4
+      - name: build
+        run: |
+          cmake -B build -DBUILD_EXAMPLES=ON -G "Visual Studio 16 2019" -A ${{matrix.arch}}
+          cmake --build build --config ${{matrix.buildType}}
```
Comment (bnoordhuis) [inline .github/workflows/ci.yml]:
```suggestion
          cmake --build build --config ${{matrix.buildType}} --target qjs_exe
          cmake --build build --config ${{matrix.buildType}} --target function_source
          cmake --build build --config ${{matrix.buildType}} --target fib
          cmake --build build --config ${{matrix.buildType}} --target point
```
i.e., omit run-test262; 2019 doesn't seem to suport `_Thread_local`

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -217,6 +217,31 @@ jobs:
...
+      - name: stats
+        run: |
+          build\${{matrix.buildType}}\qjs.exe -qd
+      - name: test
+        run: |
+          cp build\${{matrix.buildType}}\fib.dll examples\
+          cp build\${{matrix.buildType}}\point.dll examples\
+          build\${{matrix.buildType}}\qjs.exe examples\test_fib.js
+          build\${{matrix.buildType}}\qjs.exe examples\test_point.js
+          build\${{matrix.buildType}}\run-test262.exe -c tests.conf
```
Comment (bnoordhuis) [inline .github/workflows/ci.yml]:
```suggestion
```

File: quickjs-c-atomics.h [hunk 3]
```diff
@@ -49,6 +49,47 @@
...
+#define atomic_load_explicit(p_a, mo) atomic_load(p_a)
+
+static inline int atomic_compare_exchange_strong_int(LONG *obj, LONG *expected,
+                                                     LONG desired)
+{
+    LONG orig = *expected;
+    *expected = InterlockedCompareExchange(obj, desired, orig);
+    return *expected == orig;
+}
+#define atomic_compare_exchange_strong(p_a, expected, desired) atomic_compare_exchange_strong_int((LONG *)p_a, (LONG *)expected, (LONG)desired)
```
Comment (bnoordhuis) [inline quickjs-c-atomics.h]: This and atomic_load/atomic_store/etc. aren't safe. quickjs.c definitely calls them on values that aren't sizeof(LONG).

**Review by bnoordhuis — COMMENTED** [review]
The polyfill is broken. It won't work for e.g. `_Atomic uint8_t *`.

File: quickjs-c-atomics.h (hunk 3 above)
Comment (satk0) [reply quickjs-c-atomics.h]: Will removing the casting to `LONG` make them safe?

**Comment by satk0** [comment]
How come that `macos-12` is failing now all of the sudden 🤔

File: quickjs-c-atomics.h (hunk 3 above)
Comment (bnoordhuis) [reply quickjs-c-atomics.h]: No. There's https://learn.microsoft.com/en-us/previous-versions/ttk2z1ws(v=vs.85) for compare-and-swap for 16, 32 and 64 bits values, but not 8 bits, AFAIK.

Comment (satk0) [reply quickjs-c-atomics.h]: Will replacing these functions with `_` alternatives do?

Comment (bnoordhuis) [reply quickjs-c-atomics.h]: My suggestion would be to use [`_Generic`](https://en.cppreference.com/w/c/language/generic), to dispatch based on the type. That _should_ (fingers crossed) work with 2019.

**Comment by trufae** [comment]
Looks like ready to merge

**Comment by bnoordhuis** [comment]
https://github.com/quickjs-ng/quickjs/pull/592/files#r1797425774 is still unresolved.

I don't know of a way to implement compare-and-swap for 8 bit types (not without falling back to assembly) so maybe the best way forward is to:

1. disable CONFIG_ATOMICS when building for 2019 (see define in quickjs.c)
2. add `_Generic` int32/uint32 overloads for atomic_load, atomic_store and atomic_fetch_add
3. don't build run-test262 (already the case in this PR)

**Comment by saghul** [comment]
The CO does pass but I think Ben's comments haven't been addressed yet.

**Comment by saghul** [comment]
> https://github.com/quickjs-ng/quickjs/pull/592/files#r1797425774 is still unresolved.
> 
> 
> 
> I don't know of a way to implement compare-and-swap for 8 bit types (not without falling back to assembly) so maybe the best way forward is to:
> 
> 
> 
> 1. disable CONFIG_ATOMICS when building for 2019 (see define in quickjs.c)
> 
> 2. add `_Generic` int32/uint32 overloads for atomic_load, atomic_store and atomic_fetch_add
> 
> 3. don't build run-test262 (already the case in this PR)

The problem is libc now depends on atomics too (can_poll) so that's a problem...

**Comment by bnoordhuis** [comment]
Yes, that's addressed by (2), generic overloads for int32 and uint32 for three atomic operations. quickjs-libc.c doesn't use anything else (for now, anyway.)

**Comment by saghul** [comment]
Great, let's strive for 2 then!

**Comment by satk0** [comment]
Woow, didn't know that you can invoke a different function based on different type in C, will try that, thanks :sunglasses:

**Comment by satk0** [comment]
Done

EDIT: Eh, again fail :///

**Comment by saghul** [comment]
The build step looks successful but it really failed with some syntax errors...

**Comment by satk0** [comment]
`quickjs-libc.c(3847,5): error C7702: no compatible type for 'int *' in _Generic association list`, I think that's the issue but I don't know how to solve it, could u plz help?

**Comment by saghul** [comment]
I just landed the PR which removed the need for atomics on quickjs-libc. So, you can probably get away with not implementing support for them yourself now ;-)

**Comment by satk0** [comment]
Great! Thanks a lot, I will try to update qjs on radare2 and see how it goes, will let u know

File: quickjs.c [hunk 4]
```diff
@@ -62,7 +62,7 @@
 #define NO_TM_GMTOFF
 #endif
 
-#if !defined(EMSCRIPTEN) && !defined(__wasi__) && !defined(__STDC_NO_ATOMICS__)
+#if !defined(EMSCRIPTEN) && !defined(__wasi__) && !__STDC_NO_ATOMICS__
```
Comment (saghul) [inline quickjs.c:65]: We'd still need this change I think.

Comment (satk0) [reply quickjs.c:65]: oh yea, that's right

Comment (satk0) [reply quickjs.c:65]: Can I make a small PR for that or fix this one so you could merge it? @saghul

Comment (saghul) [reply quickjs.c:65]: Sure. Let's add the CI too I guess. Just build the qjs target so the test runner doesn't get built.

Comment (satk0) [reply quickjs.c:65]: Ok so I will fix this one

Comment (satk0) [reply quickjs.c:65]: I will delete all the atomics things I tried in `quickjs-c-atomics.h` as they don't work, ok?

Comment (saghul) [reply quickjs.c:65]: Sure thing!

Comment (satk0) [reply quickjs.c:65]: Seems like it's greening, is that exactly what you wanted? If not I could fix it. Nevertheless big thanks for all !

Comment (saghul) [reply quickjs.c:65]: Yep, this is it!

**Comment by satk0** [comment]
Short note: It's working now for radare2 !!

**Comment by saghul** [comment]
Great to hear! 👏👏

**Pull request merged**